typescript/packages/wallets/crossmint/src/wallets/SolanaSmartWalletClient.ts

```diff
@@ -0,0 +1,171 @@
+              userId?: number;
+          };
+          address?: never;
+      }
+    | {
+          address: string;
+          linkedUser?: never;
+      }
+);
+
+function getLocator(address: string | undefined, linkedUser: SolanaSmartWalletOptions["linkedUser"]): string {
+    if (linkedUser) {
+        if (linkedUser.email) return `email:${linkedUser.email}:solana-smart-wallet`;
+        if (linkedUser.phone) return `phone:${linkedUser.phone}:solana-smart-wallet`;
+        if (linkedUser.userId) return `userId:${linkedUser.userId}:solana-smart-wallet`;
+    }
+    if (!address) throw new Error("Either address or linkedUser must be provided");
+    return address;
+}
+
+export class SolanaSmartWalletClient extends SolanaWalletClient {
+    readonly #locator: string;
+    readonly #api: CrossmintWalletsAPI;
+    readonly #address: string;
+    readonly #adminSigner: SolanaSmartWalletOptions["config"]["adminSigner"];
+
+    constructor(api: CrossmintWalletsAPI, address: string, options: SolanaSmartWalletOptions) {
+        super({ connection: options.connection });
+        this.#api = api;
+        this.#address = address;
+        this.#adminSigner = options.config.adminSigner;
+        this.#locator = getLocator(options.address, options.linkedUser);
+    }
+
+    private static deriveAddressFromSecretKey(secretKey: string): string {
+        try {
+            const decoded = bs58.decode(secretKey);
+            const keyPair = nacl.sign.keyPair.fromSecretKey(decoded);
+            return bs58.encode(Buffer.from(keyPair.publicKey));
+        } catch (error) {
+            throw new Error(`Invalid secret key: ${error}`);
+        }
+    }
+
+    private async sendApprovals(transactionId: string, message: string, signerPrivateKey: Uint8Array): Promise<void> {
+        try {
+            const signature = nacl.sign.detached(bs58.decode(message), signerPrivateKey);
+            const encodedSignature = bs58.encode(signature);
+
+            const signerStr =
+                this.#adminSigner.type === "solana-keypair"
+                    ? `solana-keypair:${SolanaSmartWalletClient.deriveAddressFromSecretKey(this.#adminSigner.secretKey)}`
+                    : "";
+
+            // Send approval with signature
+            const approvals = [
+                {
+                    signer: signerStr,
+                    signature: encodedSignature,
+                },
+            ];
+
+            await this.#api.approveTransaction(this.#locator, transactionId, approvals);
+        } catch (error) {
+            throw new Error(`Failed to send transaction approval: ${error}`);
+        }
+    }
+
+    getAddress() {
+        return this.#address;
+    }
+
+    async signMessage(message: string): Promise<Signature> {
+        throw new Error("Signing not supported for smart wallets");
+    }
+
+    async sendTransaction({
+        instructions,
+        addressLookupTableAddresses = [],
+    }: SolanaTransaction): Promise<{ hash: string }> {
```

Comment:
Regarding the comment below, it could be
```suggestion
    async sendTransaction({
        instructions,
        addressLookupTableAddresses = [],
    }: SolanaTransaction, additionalSigners: Array<Keypair>): Promise<{ hash: string }> {
```